platform/darwin/src/MGLOfflineTask.mm

```diff
@@ -0,0 +1,189 @@
+#import "MGLOfflineTask_Private.h"
+
+#import "MGLOfflineStorage_Private.h"
+#import "MGLOfflineRegion_Private.h"
+#import "MGLTilePyramidOfflineRegion.h"
+
+#include <mbgl/storage/default_file_source.hpp>
+#include <mbgl/util/string.hpp>
+
+#define MGLAssertOfflineTaskIsValid() \
+    do { \
+        if (_state == MGLOfflineTaskStateInvalid) { \
+            [NSException raise:@"Invalid offline task" \
+                        format: \
+             @"-[MGLOfflineStorage removeTask:withCompletionHandler:] has been called " \
+             @"on this instance of MGLOfflineTask, rendering it invalid. It is an " \
+             @"error to send any message to this task."]; \
+        } \
+    } while (NO);
+
+class MBGLOfflineRegionObserver;
+
+@interface MGLOfflineTask ()
+
+@property (nonatomic, readwrite) mbgl::OfflineRegion *mbglOfflineRegion;
+@property (nonatomic, readwrite) MBGLOfflineRegionObserver *mbglOfflineRegionObserver;
+@property (nonatomic, readwrite) MGLOfflineTaskState state;
+@property (nonatomic, readwrite) MGLOfflineTaskProgress progress;
+
+@end
+
+@implementation MGLOfflineTask
+
+- (instancetype)init {
+    [NSException raise:@"Method unavailable"
+                format:
+     @"-[MGLOfflineTask init] is unavailable. "
+     @"Use +[MGLOfflineStorage addTaskForRegion:withContext:completionHandler:] instead."];
+    return nil;
+}
+
+- (instancetype)initWithMBGLRegion:(mbgl::OfflineRegion *)region {
+    if (self = [super init]) {
+        _mbglOfflineRegion = region;
+        _state = MGLOfflineTaskStateInactive;
+        _mbglOfflineRegionObserver = new MBGLOfflineRegionObserver(self);
+        
+        mbgl::DefaultFileSource *mbglFileSource = [[MGLOfflineStorage sharedOfflineStorage] mbglFileSource];
+        mbglFileSource->setOfflineRegionObserver(*_mbglOfflineRegion, std::make_unique<MBGLOfflineRegionObserver>(*_mbglOfflineRegionObserver));
+    }
+    return self;
+}
+
+- (void)dealloc {
+    delete _mbglOfflineRegionObserver;
```

Comment:
Oh, in that case, the `_mbglOfflineRegion` ivar might need to be marked `unsafe_unretained`, to keep ARC from holding its own reference to the observer.